src/cryptonote_core/service_node_rules.cpp

```diff
@@ -32,17 +32,41 @@ uint64_t portions_to_amount(uint64_t portions, uint64_t staking_requirement)
     return resultlo;
 }
 
-bool check_service_node_portions(const std::vector<uint64_t>& portions)
+bool check_service_node_portions(uint8_t hf_version, const std::vector<uint64_t>& portions)
 {
-    uint64_t portions_left = STAKING_PORTIONS;
+    if (portions.size() > MAX_NUMBER_OF_CONTRIBUTORS) return false;
 
-    for (const auto portion : portions) {
-        const uint64_t min_portions = std::min(portions_left, MIN_PORTIONS);
-        if (portion < min_portions || portion > portions_left) return false;
-        portions_left -= portion;
+    uint64_t reserved = 0;
+    for (auto i = 0u; i < portions.size(); ++i) {
+        const uint64_t min_portions = get_min_node_contribution(hf_version, STAKING_PORTIONS, reserved, i);
+        if (portions[i] < min_portions) return false;
+        reserved += portions[i];
     }
 
-    return true;
+    return reserved <= STAKING_PORTIONS;
+}
+
+
+static uint64_t get_min_node_contribution_pre_v11(uint64_t staking_requirement, uint64_t total_reserved)
+{
+    return std::min(staking_requirement - total_reserved, staking_requirement / MAX_NUMBER_OF_CONTRIBUTORS);
+}
+
+uint64_t get_min_node_contribution(uint8_t version, uint64_t staking_requirement, uint64_t total_reserved, size_t contrib_count)
+{
+    if (version < cryptonote::network_version_11_swarms) {
+        return get_min_node_contribution_pre_v11(staking_requirement, total_reserved);
+    }
+
+    const uint64_t needed = staking_requirement - total_reserved;
+    const size_t vacant = MAX_NUMBER_OF_CONTRIBUTORS - contrib_count;
+
+    /// This function is not called when the node is full, but if
+    /// it does in the future, at least we don't cause undefined behaviour
+    if (vacant == 0) return 0;
+    assert(vacant > 0); /// sanity, should never happen
```

Comment:
Huh? What's the point in that?
This function has a precondition that it should not be called when the service node is full, and we already check that before calling it.

Reply:
A precondition enforced by a comment, yes. We want the assert to fire in debug mode, if this happens in release mode the return 0 makes it have sane behaviour.

The preexisting code uses it correctly yes. Future code? Maybe.

Reply:
> This function has a precondition that it should not be called when the service node is full, and we already check that before calling it.

**Correction**: this precondition is that we don't have more contributors than we allow (not that the node is not full), which seems reasonable.

> A precondition enforced by a comment, yes.

That is the meaning of precondition, it is not enforced by code, otherwise it wouldn't be. It is like we agree that we don't commit a stake transaction unless we call `check_stake_allowed` first. That said, I agree that we should minimise those.

By the way, we disable `assert` in release mode.

Reply:
The point is, we program defensively and enforcing invariants in code is better than comments. In development we're in debug mode, which we want assertions to trigger so we fail fast and can fix it.